dtk/test/test_process.py

```diff
@@ -433,3 +433,145 @@ def test_time_vector():
 
     expected_time = [1.0, 2.0, 3.0, 4.0, 5.0]
     testing.assert_allclose(process.time_vector(5, 1.00, 1.0), expected_time)
+
+
+class TestSpectralAnalysis:
+    """
+    Test the functions freq_spectrum(), pow_spectrum() and cum_pow_spectrum()
+    based on a common test signal.
+
+    """
+    
+    def setup(self):
+        """
+        Create test signal.
+
+        """
+        # test time and sampling rate. Make sure that the number of samples is 
+        # a power of two to prevent frequency leakage (freq_spectrum does zero-
+        # padding)
+        T = 5.12      #signal period
+        self.f_s = 100     #sample rate
+        
+        t = np.arange(0, T, 1/self.f_s)
+        N = len(t)
+        k = np.arange(0, N)
+        
+        # test function
+        def harmonics(t, frequencies, amplitudes):
+            """
+            Create a test function with three harmonic frequencies.
+            
+            x(t) = A0 * sin(2*pi*f0*t) + ... + An * sin(2*pi*fn*t)
+            
+            Parameters
+            ----------
+            t : array-like
+                Array of sample times
+            frequencies : array-like
+                Array of n frequencies 
+            amplitudes : array-like
+                Array of n amplitudes 
+            
+            Returns
+            -------
+            x : array-like
+                Array of samples of the signal.
+            """
+            x = np.zeros_like(t, dtype=float)
+            for f, A in zip(frequencies, amplitudes):
+                x += A * np.sin(2*np.pi*f*t) 
+            return x
+        
+        #Build a test signal of three harmonics. The frequencies have to be 
+        #multiples of the frequency resolution to prevent frequency leakage. 
+        self.frequencies_k = np.array((5, 12, 17, 28, 25, 40)) 
+        self.frequencies = self.frequencies_k * self.f_s / N
+        self.amplitudes = np.array((0.5, 0.9, 0.4, 0.3, 0.6, 0.1))
+        
+        self.x = harmonics(t, self.frequencies, self.amplitudes)
+        
+    def test_freq_spectrum(self):
+        
+        # call frequency spectrum 
+        freq, amp = process.freq_spectrum(self.x, self.f_s) 
+        m = np.arange(0, len(freq))
+        
+        #build expected spectrum. This should have discrete peaks at the chosen
+        #frequencies with the designed amplitude of the harmonic signal frequs.
+        amp_exp = np.zeros_like(amp)
+        for fk, a in zip(self.frequencies_k, self.amplitudes):
+            amp_exp[fk-1] = a   #shift by 1, bc freq_spectrum skips f=0
+        
+        #plot the results for inspection
+        #plt.plot(freq, amp   , "r.")
+        #plt.plot(freq, amp_exp, "b.")
+        
+        assert np.allclose(amp, amp_exp), "The result freq_spectrum does n" + \
+                                          "ot match the theoretical expect" + \
+                                          "ation."
```

Comment:
It is preferable to not use backslashes these days. You can always do:

```python
msg = ("The result freq_spectrum "
      "not match the theoretical "
      "expectation.")
assert np.allclose(amp, amp_exp), msg
```

Reply:
Done!